### PR TITLE
diff/fd,optimize: don't allocate matrix on nil input to fd.Hessian and use mat.SymDense in optimize

### LINKED ISSUE #223: optimize,diff/fd: impedance mismatch in Hessian and mat.MutableSymmetric
recently, I wanted to numerically compute a hessian, so I used:
https://godoc.org/gonum.org/v1/gonum/diff/fd#Hessian
```go
// Hessian approximates the Hessian matrix of the multivariate function f
// at the location x. That is
//  H_{i,j} = ∂^2 f(x)/∂x_i ∂x_j
// If dst is not nil, the resulting H will be stored in-place into dst
// and returned, otherwise a new matrix will be allocated first. Finite difference
// formula and other options are specified by settings. If settings is nil,
// the Hessian will be estimated using the Forward formula and a default step size.
//
// Hessian panics if the size of dst and x is not equal, or if the derivative
// order of the formula is not 1.
func Hessian(
    dst *mat.SymDense, f func(x []float64) float64, 
    x []float64, settings *Settings,
) *mat.SymDense
```

so I could feed that into `optimize.Problem.Hess`:
https://godoc.org/gonum.org/v1/gonum/optimize#Problem
```go
type Problem struct {
    // Func evaluates the objective function at the given location. Func
    // must not modify x.
    Func func(x []float64) float64

    // Grad evaluates the gradient at x and stores the result in-place in grad.
    // Grad must not modify x.
    Grad func(grad []float64, x []float64)

    // Hess evaluates the Hessian at x and stores the result in-place in hess.
    // Hess must not modify x.
    Hess func(hess mat.MutableSymmetric, x []float64)

    // Status reports the status of the objective function being optimized and any
    // error. This can be used to terminate early, for example when the function is
    // not able to evaluate itself. The user can use one of the pre-provided Status
    // constants, or may call NewStatus to create a custom Status value.
    Status func() (Status, error)
}
```

so, I had to resort to:

```go
p.Hess = func(hess mat.MutableSymmetric, x []float64) {
	fd.Hessian(hess.(*mat.SymDense), f.fct, x, nil)
}
```

looking at `fd.Hessian`, it seems to me the only requirement for using `*mat.SymDense` is coming from the ability to pass `nil` as an optional parameter (as a Hessian matrix) and return it (as an output value.)

couldn't we just use `mat.MutableSymmetric` everywhere ?

### PR DESCRIPTION
**Pull request by kortschak**
Please take a look.

Closes  #223.
Updates #593.

### CODE REVIEW

File: optimize/minimize.go [hunk 1]
```diff
@@ -480,7 +480,11 @@ func evaluate(p *Problem, loc *Location, op Operation, x []float64) {
 		loc.Gradient = p.Grad(loc.Gradient, x)
 	}
 	if op&HessEvaluation != 0 {
-		loc.Hessian = p.Hess(loc.Hessian, x)
+		// Make sure we have a destination in which to place the Hessian.
```
Comment (kortschak) [inline optimize/minimize.go]: This is here because it makes the initialisation vastly more convenient at the cost of some simple pixie dust. There is no documentation for this, though it could be added to `Location`, saying that `Hessian` will be allocated if needed.

File: diff/fd/hessian.go [hunk 2]
```diff
@@ -11,28 +11,26 @@ import (
...
-	} else {
-		if n2 := dst.Symmetric(); n2 != n {
-			panic("hessian: dst size mismatch")
-		}
-		for i := 0; i < n; i++ {
-			for j := i; j < n; j++ {
-				dst.SetSym(i, j, 0)
-			}
+	if dst.IsZero() {
+		*dst = *(dst.GrowSym(n).(*mat.SymDense))
```
Comment (btracey) [inline diff/fd/hessian.go]: Outside the scope of this PR, but we should think about making an easier version of this. I'm thinking basically an exported version of `ReuseAs`. Then this can all be
```
n := len(x)
dst.ReuseAs(n)
dst.Zero()
``` 
(or ReuseAsSym). This would simplify code and make it easier for outside implementers.

File: diff/fd/hessian.go [hunk 3]
```diff
@@ -11,28 +11,26 @@ import (
...
+	} else if dst.Symmetric() != n {
+		panic("hessian: dst size mismatch")
+	}
+	for i := 0; i < n; i++ {
```
Comment (btracey) [inline diff/fd/hessian.go]: While here, just dst.Zero()?

File: optimize/minimize.go (hunk 1 above)
Comment (btracey) [reply optimize/minimize.go]: SGTM.

File: diff/fd/hessian.go (hunk 2 above)
Comment (btracey) [reply diff/fd/hessian.go]: Though, perhaps the public `ReuseAs` should always zero, because otherwise it will be a very convenient source of bugs.

File: optimize/types.go [hunk 4]
```diff
@@ -129,11 +129,9 @@ type Problem struct {
 	Grad func(grad []float64, x []float64) []float64
 
 	// Hess evaluates the Hessian at x and stores the result in-place in hess.
-	// Hess must not modify x. Hess may use (and return) the provided Symmetric
-	// if it is non-nil, or must allocate a new Symmetric otherwise. Minimize
-	// will 'give back' the returned Symmetric where possible, allowing Hess
-	// to use a type assertion on the provided matrix.
-	Hess func(hess mat.Symmetric, x []float64) mat.Symmetric
+	// Hess must not modify x. Hess may use the provided mat.SymDense, and will
```
Comment (btracey) [inline optimize/types.go]: In this case, `Hess` must use the provided `mat.SymDense`, and must resize it if zero-valued. With a TODO to document ReuseAs if/when it is added.

File: diff/fd/hessian.go (hunk 2 above)
Comment (kortschak) [reply diff/fd/hessian.go]: I think just a local helper would be appropriate.

File: diff/fd/hessian.go (hunk 3 above)
Comment (kortschak) [reply diff/fd/hessian.go]: I forgot we had that. I was wishing for one, but didn't bother looking.

File: optimize/types.go (hunk 4 above)
Comment (kortschak) [reply optimize/types.go]: Done the comment here, but I'm not sure what you want with the TODO.

**Comment by kortschak** [comment]
PTAL

I was also wanting broad agreement for this given how infrequently I come here.

File: optimize/functions/functions.go [hunk 5]
```diff
@@ -60,14 +60,13 @@ func (Beale) Grad(grad, x []float64) []float64 {
 	return grad
 }
 
-func (Beale) Hess(hess mat.Symmetric, x []float64) mat.Symmetric {
+func (Beale) Hess(hess *mat.SymDense, x []float64) {
```
Comment (vladimir-ch) [inline optimize/functions/functions.go]: Should these be `dst` instead of `hess`?

Comment (kortschak) [reply optimize/functions/functions.go]: Yes. I think they should.

Done

**Comment by kortschak** [comment]
@btracey and @vladimir-ch Are you both OK with this?

File: diff/fd/hessian.go (hunk 2 above)
Comment (btracey) [reply diff/fd/hessian.go]: I think it would be nice to make it easy, especially to encourage outside users to follow the API conventions

Comment (kortschak) [reply diff/fd/hessian.go]: Can you file a proposal issue for that?

File: optimize/types.go [hunk 6]
```diff
@@ -129,11 +129,9 @@ type Problem struct {
...
 
 	// Hess evaluates the Hessian at x and stores the result in-place in hess.
-	// Hess must not modify x. Hess may use (and return) the provided Symmetric
-	// if it is non-nil, or must allocate a new Symmetric otherwise. Minimize
-	// will 'give back' the returned Symmetric where possible, allowing Hess
-	// to use a type assertion on the provided matrix.
-	Hess func(hess mat.Symmetric, x []float64) mat.Symmetric
+	// Hess must not modify x. Hess must use the provided mat.SymDense, and
+	// must resize it if it is zero-sized.
+	Hess func(hess *mat.SymDense, x []float64)
```
Comment (vladimir-ch) [inline optimize/types.go]: I'm slightly concerned about the discord between `Grad` and `Hess` but I don't have an alternative other than Grad taking a non-nil pointer to correctly sized or zero sized slice.

File: diff/fd/hessian.go [hunk 7]
```diff
@@ -11,30 +11,24 @@ import (
...
-// the Hessian will be estimated using the Forward formula and a default step size.
+// The resulting H will be stored in dst. Finite difference formula and other
+// options are specified by settings. If settings is nil, the Hessian will be
+// estimated using the Forward formula and a default step size.
 //
-// Hessian panics if the size of dst and x is not equal, or if the derivative
-// order of the formula is not 1.
-func Hessian(dst *mat.SymDense, f func(x []float64) float64, x []float64, settings *Settings) *mat.SymDense {
+// If the dst matrix is zero-sized it will be resized to the correct dimensions,
+// otherwise dst must match the dimension of the receiver or CovarianceMatrix
```
[the remark this answers is not in the document]
Comment (kortschak) [reply diff/fd/hessian.go]: Yes. Sorry about that.

**Comment by btracey** [comment]
Dan, before submitting could you file a bug to fix up `optimize.Result`?

File: optimize/types.go (hunk 6 above)
Comment (kortschak) [reply optimize/types.go]: I've been thinking about that. The difference and justification is that `Hess` takes a pointer value destination and `Grad` takes a non-pointer destination. This means that in order to resize the `Grad` destination in a way that the caller can see, there needs to be a returned value. This doesn't need to happen with the `Hess` destination. We could have them always returning the destination, but then `Hess` has two potential sources of truth, the passed in parameter and the returned value. Since these are potentially user-implemented functions, this is a crack for bugs to live in.

In short, I'm not wildly happy with it, but I think it's the least worst.

**Comment by kortschak** [comment]
I'm not sure what it is that you want to change there.

BTW. I'm happy to fix up the copy/paste error that @vladimir-ch found. Given that you all have approved the substance of the PR, I'd only need a single approver after that.

**Comment by kortschak** [comment]
PTAL

**Comment by kortschak** [comment]
@vladimir-ch Sorry, I had thought that you had approved. I'll wait for your approval.

File: optimize/types.go (hunk 6 above)
Comment (btracey) [reply optimize/types.go]: If we are doing it this way, we could go back to the old `optimize` way, where the user didn't need to care about it (and had to store in-place). We do need the ability for the `Method` to set the value to `nil`, but that case could checked and allocated in `Minimize`.  This was not possible for an arbitrary `Symmetric` but is very easy with a fixed `*SymDense`.

Comment (btracey) [reply optimize/types.go]: This would be simpler for users and would still mesh with `fd`, since the matrix would always be allocated.

Comment (kortschak) [reply optimize/types.go]: I'm unclear what the old `optimize` way was. What was the signature?

Comment (btracey) [reply optimize/types.go]:
```
Grad: func(dst, x []float64)
Hess: func(dst *mat.SymDense, x []float64)
```

The difference being that `Minimize` guarantees that `len(dst) == len(x)` in the call to grad, and `dst.Symmetric() == len(x)` in the call to `Hess`. In other words, `Minimize` guarantees they are the right size and the user doesn't have to worry about checking if they are nil or not.

Comment (kortschak) [reply optimize/types.go]: Ah. OK. Would you like me to update this for that behaviour?

Comment (btracey) [reply optimize/types.go]: I think so, yes. That check should probably happen in `evaluate`, minimize.go ~ line 470 (and there would be corresponding documentation changes in `Problem`)

Comment (kortschak) [reply optimize/types.go]: There's a TODO for consideration. I can do that in this PR if you want; whether the allocation of the gradient slice is when it is nil, or when it is zero-length.

Comment (btracey) [reply optimize/types.go]: We can leave as a TODO. I suspect the correct behavior is to only check if the relevant entry is `nil`. These values are never allocated by the provider of `Problem`, this is only for communicating between `Minimize` and `Method`. The reason to allow nil values is so the `Method` can safely save the values. It seems like allowing zero values as well could be ripe for bugs, in that it's mid-way between explicitly keeping the data and explicitly destroying it.

Comment (kortschak) [reply optimize/types.go]: That would suggest that the Hessian should also only be a nil check then.

**Comment by btracey** [comment]
Agreed. Sorry, I should have been more clear

File: optimize/types.go (hunk 6 above)
Comment (kortschak) [reply optimize/types.go]: Done

**Comment by kortschak** [comment]
PTAL

**Comment by kortschak** [comment]
It feels a lot cleaner now.

I have to say, doing this in, say python, would have been horrible.

File: optimize/types.go (hunk 6 above)
Comment (vladimir-ch) [reply optimize/types.go]: This is better.

**Comment by kortschak** [comment]
@sbinet Are you happy with this?
